crates/query-engine/translation/src/translation/query/fields.rs

```diff
@@ -318,3 +323,135 @@ fn translate_nested_field(
         },
     ))
 }
+
+#[allow(clippy::too_many_arguments)]
+/// In order to return the expected type representation for each column,
+/// we need to wrap columns in type representation cast, and unpack composite types
+/// so we can wrap them.
+fn unpack_and_wrap_fields(
+    env: &Env,
+    state: &mut State,
+    current_table: &TableNameAndReference,
+    join_relationship_fields: &mut Vec<relationships::JoinFieldInfo>,
+    column: &str,
+    alias: sql::ast::ColumnAlias,
+    type_info: &CollectionOrCompositeTypeInfo<'_>,
+    nested_field_joins: &mut Vec<JoinNestedFieldInfo>,
+) -> Result<(sql::ast::ColumnAlias, sql::ast::Expression), Error> {
+    let column_info = type_info.lookup_column(column)?;
+
+    // Different kinds of types have different strategy for converting to their
+    // type representation.
+    match column_info.r#type {
+        // Scalar types can just be wrapped in a cast.
+        Type::ScalarType(_) => {
+            let column_type_representation = env.lookup_type_representation(&column_info.r#type);
+            let (alias, expression) = sql::helpers::make_column(
+                current_table.reference.clone(),
+                column_info.name.clone(),
+                alias,
+            );
+            Ok((
+                alias,
+                wrap_in_type_representation(expression, column_type_representation),
+            ))
+        }
+        // Composite types are a more involved case because we cannot just "cast"
+        // a composite type, we need to unpack it and cast the individual fields.
+        // In this case, we unpack a single composite column selection to an explicit
+        // selection of all fields.
+        Type::CompositeType(ref composite_type) => {
+            // build a nested field selection of all fields.
+            let nested_field = models::NestedField::Object({
+                let composite_type = env.lookup_composite_type(composite_type)?;
+                let mut fields = indexmap!();
+                for (name, field_info) in composite_type.info.fields.iter() {
+                    fields.insert(
+                        name.to_string(),
+                        models::Field::Column {
+                            column: field_info.name.clone(),
+                            fields: None,
+                        },
+                    );
```

Comment:
Ah, I think this is the root of our CompositeTypeInfo/CollectionOrCompositeTypeInfo-related misunderstanding.

It's perfectly possible for our `composite_type` bound in `Type::CompositeType(ref composite_type)` to come from both tables _and_ composite types.